### DISCUSSION TITLE
Ability to write Skins in C# like Argon / Triangles / Argon Pro is

### DISCUSSION THREAD

**Discussion by Lauriichan** [ppy/osu#29718] [Ideas] [answered]
Right now the only documented way to add custom skins to lazer is using the legacy skin format.
However if you look at elements that the Triangles skin adds for example the hit circles have the same animation as buttons and the main menu, they can just render more than legacy skins.
Since skins are already loaded dynamically the only part missing right now would be to have a dll loader for the C# skins.
If don't know if it is already possible without forking the client and modifying it directly or not so that's why I'm opening this right here.

**Comment by Alib234** [comment]
I could see this being abused a lot since it would basically allow people to write any arbitrary code and have the game execute it while being able to submit scores, ex. cheats, exploits etc.
This would probably be impossible to account for and detect since I'm sure someone could bypass the checks in one way or another.
Only way I really could see this happen is if there was some integrated skin list thing in lazer for those types of skins, with all of them checked beforehand by someone to see if they contain malicious code or not.

**Comment by peppy** [comment, marked as answer]
While this is out of scope of the game, we're focused on providing more settings and options for skinning. But if required, you could PR the elements you need for your skin, and then use them once merged. In the future you will be able to use white-listed drawables to make up skin components.

**Reply by Lauriichan** [reply]
So will there be a new skin format later on or will there only be legacy skins?

**Reply by peppy** [reply]
There already is.

**Reply by Lauriichan** [reply]
When I was searching for documentation on how to create custom skins expect for the legacy skin format I couldn't find anything, so this is not publicly available yet?
Or is it available but I'm just too stupid to find the documentation, if that is the case would it be possible that you could link me to it, thank you for all your responses so far.

**Reply by peppy** [reply]
There's no documentation as it is intended to be managed from the game's UI only.
